python/lsst/resources/http.py

```diff
@@ -371,18 +385,85 @@ def is_webdav_endpoint(self) -> bool:
+            # request, even if the behavior for such a request against a
+            # directory is not specified, so it depends on the server
+            # implementation.
+            resp = self.session.head(self.geturl(), timeout=TIMEOUT, allow_redirects=True)
+            return resp.status_code == requests.codes.ok  # 200
+
+        # The remote endpoint is a webDAV server: send a PROPFIND request
+        # requesting only the 'getlastmodified' property.
+        request_body = """
+        <?xml version="1.0" encoding="utf-8" ?>
+        <D:propfind xmlns:D="DAV:"><D:prop><D:getlastmodified/></D:prop></D:propfind>
+        """
+        resp = self._propfind(request_body)
+        if resp.status_code == requests.codes.multi_status:  # 207
+            # Parse the XML-encoded response. Since we issue a request for a
+            # single remote resource, the response must include properties for
+            # that single resource. So either its response status is OK or not.
+            tree = eTree.fromstring(resp.content.strip())
+            element = tree.find("./{DAV:}response/{DAV:}propstat/[{DAV:}status='HTTP/1.1 200 OK']")
+            return element is not None
+        elif resp.status_code == requests.codes.not_found:  # 404
+            return False
+        else:
+            raise ValueError(
+                f"Unexpected status received for PROPFIND request for {self}: {resp.status_code}"
+            )
 
     def size(self) -> int:
         """Return the size of the remote resource in bytes."""
         if self.dirLike:
             return 0
 
-        resp = self.session.head(self.geturl(), timeout=TIMEOUT, allow_redirects=True)
-        if resp.status_code != 200:
-            raise FileNotFoundError(f"Resource {self} does not exist")
-        return int(resp.headers["Content-Length"])
+        if not self.is_webdav_endpoint:
+            # The remote is a plain HTTP server. Send a HEAD request to
+            # retrieve the size of the resource.
+            resp = self.session.head(self.geturl(), timeout=TIMEOUT, allow_redirects=True)
+            if resp.status_code == requests.codes.ok:  # 200
+                if "Content-Length" in resp.headers:
+                    return int(resp.headers["Content-Length"])
+                else:
+                    raise ValueError(
+                        f"Response to HEAD request to {self} does not contain 'Content-Length' header"
+                    )
+            elif resp.status_code == requests.codes.not_found:
+                raise FileNotFoundError(f"Resource {self} does not exist, status code: {resp.status_code}")
+            else:
+                raise ValueError(
+                    f"Unexpected response for HEAD request for {self}, status code: {resp.status_code}"
+                )
+
+        # The remote is a webDAV server: send a PROPFIND request to retrieve
+        # the 'getcontentlength' property of the resource.
+        request_body = """
```

Comment:
This code is non-trivial and seems to be substantially identical to the getlastmodified implementation above. Can you consider putting this in a helper function so we only have the logic once? It seems like the code from assigning to request_body to the checks on the first `element` are the same so I think that a helper method that returned the `element` for later parsing should work okay.

Reply:
Refactored. Introduced helper class and helper function to parse `PROPFIND` responses.